### PR TITLE
Add DownloadButton component

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Related to #149 by @zackkrida 

## Description
Adds a new DownloadButton component based on the DropdownButton component.

## Technical details
Adds the `filesize` dependency which is a function for displaying filesizes with their units appropriatly. It's pretty small: https://bundlephobia.com/package/filesize

## Tests
Most of the behavior of this component is already tested in the DropdownButton component's tests. The behavior around the localStorage would be worth testing though, so I'll add some unit tests for that (assuming JSDOM supports it)

To test this locally run `npm run storybook` and navigate to `/?path=/story/components-downloadbutton--default` to test the component manually. It should even let you download the files 🙂 

## Screenshots
Captura de Tela 2021-09-14 às 09 03 12

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by zackkrida** [comment]
This is looking great 🤩 
One item of interest, there's a small localStorage wrapper in https://github.com/WordPress/openverse-frontend/blob/main/src/utils/local.js you can use, with `local.get(key)` and `local.set(key, value)` methods. This exists to include a simple `window` check in SSR scenarios.

**Comment by sarayourfriend** [comment]
Awesome, thanks for the heads up @zackkrida!

File: test/unit/specs/components/download-button.spec.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  await DownloadButton.fetch.call(wrapper.vm)
+  return wrapper
+}
+
+describe('DownloadButton', () => {
+  beforeEach(() => {
+    axios.head.mockReset()
+  })
+
+  // TODO(@sarayourfriend) convert this to testing-library once it's possible to get the `vm` from the wrapper
```
Comment (sarayourfriend) [inline test/unit/specs/components/download-button.spec.js:54]: This is blocked by this PR https://github.com/testing-library/vue-testing-library/pull/250

File: src/components/DownloadButton.vue [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import local from '~/utils/local'
+import DropdownButton from '~/components/DropdownButton'
+
+const LS_DOWNLOAD_FORMAT_EXTENSION_KEY = 'openverse:download-format-extension'
+
+export default {
+  name: 'DownloadButton',
+  components: { DropdownButton },
+  props: {
+    formats: {
```
Comment (obulat) [inline src/components/DownloadButton.vue:68]: Very thorough validation!

File: src/components/DownloadButton.vue [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+          <button
+            class="flex justify-between w-full"
+            :class="[
+              itemClass,
+              selectedFormat.extension_name === format.extension_name
+                ? activeItemClass
+                : '',
+            ]"
+            type="button"
+            v-bind="itemA11yProps"
```
Comment (obulat) [inline src/components/DownloadButton.vue:40]: Can we change the `aria-label` from dropdown to something like 'Select the download file format'?

**Comment by obulat** [comment]
The button works and looks amazing! I have a couple of questions about functionality. 
When you open a dropdown and select the filetype, shouldn't the dropdown close after selection? That was my expectation.
When using keyboard to navigate, do you click on <kbd>Shift</kbd><kbd>Tab</kbd> to go back from the dropdown to download button?

**Review by obulat — APPROVED** [review]

File: src/components/DownloadButton.vue [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    return { selectedFormat: format, filesizes: null }
+  },
+  async fetch() {
+    const extensionsToFilesizes = await Promise.all(
+      this.formats.map(async (format) => {
+        try {
+          const response = await axios.head(format.download_url)
+          return [format.extension_name, response.headers['content-length']]
+        } catch (e) {
+          return [format.extension_name, undefined]
```
Comment (zackkrida) [inline src/components/DownloadButton.vue:99]: Minor point but I love seeing undefined in use 😄

**Comment by sarayourfriend** [comment]
> When you open a dropdown and select the filetype, shouldn't the dropdown close after selection?

It should be doing that, if it's not it's a bug 😅 

> When using keyboard to navigate, do you click on ShiftTab to go back from the dropdown to download button?

Yup! I think that matches the behavior on the WARIA examples but I'll double check.

File: src/components/DownloadButton.vue [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      local.get(LS_DOWNLOAD_FORMAT_EXTENSION_KEY) ?? null
+    let format = this.formats[0]
+    if (savedFormatExtension) {
+      const found = this.formats.find(
+        (format) => format.extension_name === savedFormatExtension
+      )
+      if (found) {
+        format = found
+      }
+    }
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply src/components/DownloadButton.vue:88]: Oh huh, what's the advantage to do it that way?

[the remark this answers is not in the document]
Comment (zackkrida) [reply src/components/DownloadButton.vue:88]: I tried to research a bit and I can't find anything, so you can leave this as is, fwiw the execution order is:

1. `data()`
2. `created()`
3. `mounted()`

**Comment by sarayourfriend** [comment]
> I think there's still an issue with events firing multiple times, perhaps due to some event bubbling. Keyboard controls work flawlessly but clicks seem unpredictable. This video isn't the most helpful but should show some instances where 2,3, or even 5 clicks are necessary to reveal the dropdown.

For some reason I'm not able to reproduce this... or the bug where clicking doesn't cause the dropdown to open 🤔 I'll keep digging and see if I can identify the problem though.

**Comment by sarayourfriend** [comment]
So I've just done some experimenting and discovered something interesting... I wonder if it's at the root cause of what's going on here. When you `@click` bind an event in Vue on a `button`, it will also trigger the click when you hit space or enter... So if you combine `@click` with `@keydown` it'll fire BOTH events at the same time! You can see this happening here in this codesandbox:

https://codesandbox.io/s/reverent-lucy-j7kf3?file=/src/App.vue

I have absolutely no idea how to circumvent this... I suppose we could just not use the `button` element and use a `div` with some appropriate `aria` attributes applied... but is that the right solution? Or is there a different vue-y way to do this?

**Comment by zackkrida** [comment]
CleanShot 2021-09-23 at 13 28 20@2x

Also, small Safari bug where the button needs `appearance-none`

**Comment by zackkrida** [comment]
@sarayourfriend perhaps we just remove the keydown events? Is that all we need?

**Comment by sarayourfriend** [comment]
> perhaps we just remove the keydown events? Is that all we need?

We can do that for the dropdown button I think... but for the format buttons I don't think it's possible because we need to respond to specific keys (tab, arrows, page down/up, end/home, etc) as well as the click event... I wonder if there's a way of telling the click event to only fire if it's a true click, given we're handling the keydown event ourselves 🤔

**Comment by sarayourfriend** [comment]
Apparently you can use `https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/detail` on the `MouseEvent` to tell whether it was triggered by a `click` or a keypress 🤔

**Comment by sarayourfriend** [comment]
Hmmm, actually we can't do that because we are depending on the `@click` behavior to handle when the dropdown item is actually selected... `onItemKeydown` only handles navigational keys, not selection (which makes sense because the `DropdownButton` shouldn't know anything outside of the a11y interactions for each item...)

**Comment by sarayourfriend** [comment]
Additionally, I'm just noticing that because the link for the download button looks like a button, it doesn't actually behave like a button (<kbd>Space</kbd> doesn't trigger the download, in fact it scrolls the screen) so I wonder if there's something we need to do there to handle that situation?

**Comment by zackkrida** [comment]
> Additionally, I'm just noticing that because the link for the download button looks like a button, it doesn't actually behave like a button (Space doesn't trigger the download, in fact it scrolls the screen) so I wonder if there's something we need to do there to handle that situation?

@sarayourfriend I think it's okay for it to behave like a link and look like a button.

**Comment by sarayourfriend** [comment]
I think I figured out the issue with clicking the dropdown button. The click only triggers if you click the button and not the icon 🤯

**Comment by zackkrida** [comment]
@sarayourfriend `pointer-events: none` on the icon then should do it!

**Comment by sarayourfriend** [comment]
Thanks @zackkrida !

**Comment by sarayourfriend** [comment]
> The dropdown does not close after selecting an item, again I think because of an event double fire

I still can't figure this out... or reproduce it 😞

**Review by zackkrida — APPROVED** [review]
This now looks and works perfectly 🥳. The only bit of feedback I have is that I'd like to figure out how to show the focus ring on dropdown items only when the user is using the keyboard. If someone is using the mouse with the dropdown, it's confusing to see two interaction states: the gray background of the selected item and the pink ring defaulted to the first item.

**Pull request merged**